go/test/endtoend/vreplication/fk_ext_test.go

```diff
@@ -0,0 +1,444 @@
+	}
+	err := vc.VtctldClient.ExecuteCommand("RebuildVSchemaGraph")
+	require.NoError(t, err)
+	require.NoError(t, waitForColumn(t, vtgate, keyspaceName, "parent", "id"))
+	require.NoError(t, waitForColumn(t, vtgate, keyspaceName, "child", "parent_id"))
+	return tablets
+}
+
+func doMoveTables(t *testing.T, sourceKeyspace, targetKeyspace, workflowName, tabletTypes string, targetTabs map[string]*cluster.VttabletProcess, atomicCopy bool) {
+	mt := newMoveTables(vc, &moveTablesWorkflow{
+		workflowInfo: &workflowInfo{
+			vc:             vc,
+			workflowName:   workflowName,
+			targetKeyspace: targetKeyspace,
+			tabletTypes:    tabletTypes,
+		},
+		sourceKeyspace: sourceKeyspace,
+		atomicCopy:     atomicCopy,
+	}, workflowFlavorRandom)
+	mt.Create()
+
+	waitForWorkflowState(t, vc, fmt.Sprintf("%s.%s", targetKeyspace, workflowName), binlogdatapb.VReplicationWorkflowState_Running.String())
+
+	for _, targetTab := range targetTabs {
+		catchup(t, targetTab, workflowName, "MoveTables")
+	}
+	vdiff(t, targetKeyspace, workflowName, fkextConfig.cell, false, true, nil)
+	lg.Stop()
+	lg.SetDBStrategy("vtgate", targetKeyspace)
+	if lg.Start() != nil {
+		t.Fatal("Start failed")
+	}
+
+	mt.SwitchReadsAndWrites()
+
+	if lg.WaitForAdditionalRows(100) != nil {
+		t.Fatal("WaitForAdditionalRows failed")
+	}
+
+	waitForLowLag(t, sourceKeyspace, workflowName+"_reverse")
+	vdiff(t, sourceKeyspace, workflowName+"_reverse", fkextConfig.cell, false, true, nil)
+	if lg.WaitForAdditionalRows(100) != nil {
+		t.Fatal("WaitForAdditionalRows failed")
+	}
+
+	mt.ReverseReadsAndWrites()
+	if lg.WaitForAdditionalRows(100) != nil {
+		t.Fatal("WaitForAdditionalRows failed")
+	}
+	waitForLowLag(t, targetKeyspace, workflowName)
+	time.Sleep(5 * time.Second)
+	vdiff(t, targetKeyspace, workflowName, fkextConfig.cell, false, true, nil)
+	lg.Stop()
+	mt.SwitchReadsAndWrites()
+	mt.Complete()
+	if err := vc.VtctldClient.ExecuteCommand("ApplyRoutingRules", "--rules={}"); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func importIntoVitess(t *testing.T) {
+	targetTabs := newKeyspace(t, fkextConfig.target1KeyspaceName, "0", FKExtTarget1VSchema, FKExtSourceSchema, 200, 1)
+	doMoveTables(t, fkextConfig.sourceKeyspaceName, fkextConfig.target1KeyspaceName, "import", "primary", targetTabs, true)
+}
+
+const getConstraintsQuery = `
+SELECT CONSTRAINT_NAME, TABLE_NAME 
+FROM INFORMATION_SCHEMA.KEY_COLUMN_USAGE 
+WHERE TABLE_SCHEMA = ` + "'%s'" + ` AND REFERENCED_TABLE_NAME IS NOT NULL;
+`
+
+// dropReplicaConstraints drops all foreign key constraints on replica tables for a given keyspace/shard.
+// We do this so that we can replay binlogs from a replica which is not doing cascades but just replaying
+// the binlogs created by the primary. This will confirm that vtgate is doing the cascades correctly.
+func dropReplicaConstraints(t *testing.T, keyspaceName string, tablet *cluster.VttabletProcess) {
+	var dropConstraints []string
+
```

Comment:
Let's `require.Equal(...)` that the tablet type is `REPLICA` (or non-primary).

Reply:
Done